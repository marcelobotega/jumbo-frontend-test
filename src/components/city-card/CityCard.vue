<template>
  <div class="city-card card">
    <div
      class="card-content cursor-pointer flex justify-between"
      @click="toggleOpen()"
    >
      <p>{{ city }}</p>
      <div>
        <svg
          v-if="!isOpen"
          xmlns="http://www.w3.org/2000/svg"
          class="h-6 w-6 text-gray-400"
          fill="none"
          viewBox="0 0 24 24"
          stroke="currentColor"
        >
          <path
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            d="M19 9l-7 7-7-7"
          />
        </svg>
        <svg
          v-if="isOpen"
          xmlns="http://www.w3.org/2000/svg"
          class="h-6 w-6 text-gray-400"
          fill="none"
          viewBox="0 0 24 24"
          stroke="currentColor"
        >
          <path
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            d="M5 15l7-7 7 7"
          />
        </svg>
      </div>
    </div>
    <hr v-show="isOpen" />
    <div class="city-card-expansion" v-show="isOpen">
      <p class="text-sm font-semibold mt-1">Stores in city:</p>
      <StoreCard
        v-for="store in stores"
        :key="store.uuid"
        class="mt-3"
        :store="store"
      />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from "vue-property-decorator";
import storeService from "@/services/store.service";
import { IStore } from "@/domain/store";
import StoreCard from "@/components/store-card/StoreCard.vue";

@Component({
  components: { StoreCard },
})
export default class CityCard extends Vue {
  @Prop() city!: string;
  stores: IStore[] = [];
  isOpen = false;

  toggleOpen(): void {
    if (this.stores.length === 0) {
      this.stores = storeService.getStoresByCity(this.city);
    }
    this.isOpen = !this.isOpen;
  }
}
</script>
<style scoped lang="scss">
.city-card {
  .city-card-expansion {
    @apply m-3;
  }
}
</style>
