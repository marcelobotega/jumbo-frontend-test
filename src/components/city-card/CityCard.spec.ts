import { shallowMount } from "@vue/test-utils";
import CityCard from "@/components/city-card/CityCard.vue";
import storeService from "@/services/store.service";

describe("CityCard.vue", () => {
  it("should create", () => {
    const wrapper = shallowMount(CityCard, {
      propsData: { city: "amstelveen" },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("should open expansion panel", async () => {
    const wrapper = shallowMount(CityCard, {
      propsData: { city: "amstelveen" },
    });

    const spy = jest.spyOn(storeService, "getStoresByCity");
    await wrapper.find(".card-content").trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper).toMatchSnapshot();

    await wrapper.find(".card-content").trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper).toMatchSnapshot();

    await wrapper.find(".card-content").trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper).toMatchSnapshot();
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith("amstelveen");
  });
});
