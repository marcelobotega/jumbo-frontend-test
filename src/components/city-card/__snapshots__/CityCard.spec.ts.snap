// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityCard.vue should create 1`] = `
<div class="city-card card">
  <div class="card-content cursor-pointer flex justify-between">
    <p>amstelveen</p>
    <div><svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor" class="h-6 w-6 text-gray-400">
        <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M19 9l-7 7-7-7"></path>
      </svg>
      <!---->
    </div>
  </div>
  <hr style="display: none;">
  <div class="city-card-expansion" style="display: none;">
    <p class="text-sm font-semibold mt-1">Stores in city:</p>
  </div>
</div>
`;

exports[`CityCard.vue should open expansion panel 1`] = `
<div class="city-card card">
  <div class="card-content cursor-pointer flex justify-between">
    <p>amstelveen</p>
    <div>
      <!----> <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor" class="h-6 w-6 text-gray-400">
        <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M5 15l7-7 7 7"></path>
      </svg>
    </div>
  </div>
  <hr style="">
  <div class="city-card-expansion" style="">
    <p class="text-sm font-semibold mt-1">Stores in city:</p>
    <storecard-stub store="[object Object]" class="mt-3"></storecard-stub>
    <storecard-stub store="[object Object]" class="mt-3"></storecard-stub>
  </div>
</div>
`;

exports[`CityCard.vue should open expansion panel 2`] = `
<div class="city-card card">
  <div class="card-content cursor-pointer flex justify-between">
    <p>amstelveen</p>
    <div><svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor" class="h-6 w-6 text-gray-400">
        <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M19 9l-7 7-7-7"></path>
      </svg>
      <!---->
    </div>
  </div>
  <hr style="display: none;">
  <div class="city-card-expansion" style="display: none;">
    <p class="text-sm font-semibold mt-1">Stores in city:</p>
    <storecard-stub store="[object Object]" class="mt-3"></storecard-stub>
    <storecard-stub store="[object Object]" class="mt-3"></storecard-stub>
  </div>
</div>
`;

exports[`CityCard.vue should open expansion panel 3`] = `
<div class="city-card card">
  <div class="card-content cursor-pointer flex justify-between">
    <p>amstelveen</p>
    <div>
      <!----> <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor" class="h-6 w-6 text-gray-400">
        <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M5 15l7-7 7 7"></path>
      </svg>
    </div>
  </div>
  <hr style="">
  <div class="city-card-expansion" style="">
    <p class="text-sm font-semibold mt-1">Stores in city:</p>
    <storecard-stub store="[object Object]" class="mt-3"></storecard-stub>
    <storecard-stub store="[object Object]" class="mt-3"></storecard-stub>
  </div>
</div>
`;
